File: naddu/naddu_2/src/tc.js
class TimeControl {

  constructor() {
    this.bestMove = 0;
    this.nodes = 0;
    this.maxNodes = 0;
    this.maxDepth = 0;
    this.startTime = 0;
    this.finishTime = 0;
    this.finished = 0;
  }
}

const timeControl = new TimeControl();

function tcClear() {
  const tc = timeControl;
  tc.bestMove = 0;
  tc.nodes = 0;
  tc.maxNodes = 0;
  tc.maxDepth = 0;
  tc.startTime = now() | 0;
  tc.finishTime = 0;
  tc.finished = 0;
}

function tcCheck() {
  const tc = timeControl;
  if (tc.finishTime && now() >= tc.finishTime)
    tc.finished = 1;
  else if (tc.maxNodes && tc.nodes >= tc.maxNodes)
    tc.finished = 1;
}

function tcInit(tokens) {

  tcClear();

  const tc = timeControl;

  // Parse go command parameters
  let wtime = 0;
  let btime = 0;
  let winc = 0;
  let binc = 0;
  let movestogo = 30; // Default to 30 moves if not specified
  let movetime = 0;
  let infinite = false;

  // Parse tokens
  for (let i = 1; i < tokens.length; i++) {
    const token = tokens[i];

    switch (token) {
      case 'wtime':
        wtime = parseInt(tokens[++i]) || 0;
        break;
      case 'btime':
        btime = parseInt(tokens[++i]) || 0;
        break;
      case 'winc':
        winc = parseInt(tokens[++i]) || 0;
        break;
      case 'binc':
        binc = parseInt(tokens[++i]) || 0;
        break;
      case 'movestogo':
        movestogo = Math.max(2,parseInt(tokens[++i]) || 30);
        break;
      case 'depth':
      case 'd':
        tc.maxDepth = parseInt(tokens[++i]) || 0;
        break;
      case 'nodes':
      case 'n':
        tc.maxNodes = parseInt(tokens[++i]) || 0;
        break;
      case 'movetime':
      case 'm':
        movetime = parseInt(tokens[++i]) || 0;
        break;
      case 'infinite':
      case 'i':
        infinite = true;
        break;
    }
  }

  // Calculate finish time based on time control
  if (movetime > 0) {
    // Fixed time per move
    tc.finishTime = tc.startTime + movetime;
  }
  else if (!infinite && (wtime > 0 || btime > 0)) {
    // Calculate time allocation based on side to move
    const pos = nodes[0].pos;
    const isWhite = pos.stm === WHITE;
    const timeLeft = isWhite ? wtime : btime;
    const increment = isWhite ? winc : binc;

    // Time allocation: timeLeft / movestogo + increment / 2
    const allocatedTime = (timeLeft / movestogo) + (increment / 2);
    tc.finishTime = tc.startTime + allocatedTime;
  }

  // Set default max depth if not specified
  if (tc.maxDepth <= 0)
    tc.maxDepth = MAX_PLY;
  else if (tc.maxDepth > MAX_PLY)
    tc.maxDepth = MAX_PLY;

}

File: naddu/naddu_2/c/tc.c
#include <string.h>
#include <stdlib.h>
#include "tc.h"
#include "node.h"

#ifdef _WIN32
#include <windows.h>
#else
#include <sys/time.h>
#endif

TimeControl tc;

uint64_t now_ms(void) {
#ifdef _WIN32
  return GetTickCount64();
#else
  struct timeval tv;
  gettimeofday(&tv, NULL);
  return (uint64_t)tv.tv_sec * 1000 + tv.tv_usec / 1000;
#endif
}

void tc_clear(void) {
  tc.best_move = 0;
  tc.nodes = 0;
  tc.max_nodes = 0;
  tc.max_depth = MAX_PLY;
  tc.start_time = now_ms();
  tc.finish_time = 0;
  tc.finished = 0;
  tc.stop = 0;
}

void tc_check(void) {
  if (tc.stop) {
    tc.finished = 1;
  }
  else if (tc.finish_time && now_ms() >= tc.finish_time) {
    tc.finished = 1;
  }
  else if (tc.max_nodes && tc.nodes >= tc.max_nodes) {
    tc.finished = 1;
  }
}

void tc_init(int argc, char **tokens) {
  tc_clear();

  int wtime = 0;
  int btime = 0;
  int winc = 0;
  int binc = 0;
  int movestogo = 30;
  int movetime = 0;
  int infinite = 0;

  for (int i = 1; i < argc; i++) {
    const char *token = tokens[i];

    if (strcmp(token, "wtime") == 0 && i + 1 < argc) {
      wtime = atoi(tokens[++i]);
    }
    else if (strcmp(token, "btime") == 0 && i + 1 < argc) {
      btime = atoi(tokens[++i]);
    }
    else if (strcmp(token, "winc") == 0 && i + 1 < argc) {
      winc = atoi(tokens[++i]);
    }
    else if (strcmp(token, "binc") == 0 && i + 1 < argc) {
      binc = atoi(tokens[++i]);
    }
    else if (strcmp(token, "movestogo") == 0 && i + 1 < argc) {
      movestogo = atoi(tokens[++i]);
      if (movestogo < 2) movestogo = 2;
    }
    else if ((strcmp(token, "depth") == 0 || strcmp(token, "d") == 0) && i + 1 < argc) {
      tc.max_depth = atoi(tokens[++i]);
    }
    else if ((strcmp(token, "nodes") == 0 || strcmp(token, "n") == 0) && i + 1 < argc) {
      tc.max_nodes = atoi(tokens[++i]);
    }
    else if ((strcmp(token, "movetime") == 0 || strcmp(token, "m") == 0) && i + 1 < argc) {
      movetime = atoi(tokens[++i]);
    }
    else if (strcmp(token, "infinite") == 0 || strcmp(token, "i") == 0) {
      infinite = 1;
    }
  }

  if (movetime > 0) {
    tc.finish_time = tc.start_time + movetime;
  }
  else if (!infinite && (wtime > 0 || btime > 0)) {
    Pos *pos = &nodes[0].pos;
    int is_white = (pos->stm == WHITE);
    int time_left = is_white ? wtime : btime;
    int increment = is_white ? winc : binc;

    int allocated_time = (time_left / movestogo) + (increment / 2);
    tc.finish_time = tc.start_time + allocated_time;
  }

  if (tc.max_depth <= 0)
    tc.max_depth = MAX_PLY;
  else if (tc.max_depth > MAX_PLY)
    tc.max_depth = MAX_PLY;
}
